{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import Counter \n",
    "import re\n",
    "import itertools\n",
    "import fnmatch\n",
    "\n",
    "main_dir = \"/Users/sigalova/Desktop/Chlamydia_pangenome/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_prot2og_dict(OG_table, genome_ids):\n",
    "\n",
    "    prot2og = dict()\n",
    "    og2prot = dict()\n",
    "\n",
    "    for line in OG_table:\n",
    "        [group,proteins] = line.rstrip().split(\":\")\n",
    "        proteins = proteins.strip().split(\" \")\n",
    "        proteins = [x.split(\"|\")[1] for x in proteins if x.split(\"|\")[0] in genome_ids]\n",
    "        og2prot[group] = proteins\n",
    "        for prot in proteins:\n",
    "            prot2og[prot] = group\n",
    "            \n",
    "    return [prot2og, og2prot]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. List of genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = main_dir + \"info/genome_info_combined.csv\"\n",
    "genomes = pd.read_csv(path)\n",
    "genomes.index = genomes[\"genome_id\"]\n",
    "genome_ids_withOutgroup = genomes[\"genome_id\"].tolist() \n",
    "genome_ids_withoutOutgroup = [x for x in genome_ids_withOutgroup if x != \"W001\"]\n",
    "genome_ids_nodraft = genomes[(genomes[\"n_contigs\"]==1) & (genomes[\"genome_id\"]!=\"W001\")][\"genome_id\"].tolist()    # without outgroup, no draft genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Dict: protein_id to OG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2.A. only genus Chlamydia (with draft genomes)  \n",
    "OGs_withoutOutgroup = main_dir + '/data/ogroups2019/raw/groups_wo_outgroup_with_singletons.txt'\n",
    "\n",
    "ogs = open(OGs_withoutOutgroup, \"r\")\n",
    "res = get_prot2og_dict(ogs, genome_ids_withoutOutgroup)\n",
    "prot2og_withoutOutgroup = res[0]\n",
    "og2prot_withoutOutgroup = res[1]\n",
    "\n",
    "# 2.B. only genus Chlamydia (without draft genomes)  \n",
    "ogs = open(OGs_withoutOutgroup, \"r\")\n",
    "res = get_prot2og_dict(ogs, genome_ids_nodraft)\n",
    "prot2og_nodraft = res[0]\n",
    "og2prot_nodraft = res[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2.C. genus Chlamydia + outgroup (with draft genomes)  \n",
    "OGs_withOutgroup = main_dir + '/data/ogroups2019/raw/groups_with_outgroup_with_singletons.txt'\n",
    "\n",
    "ogs = open(OGs_withOutgroup, \"r\")\n",
    "res = get_prot2og_dict(ogs, genome_ids_withOutgroup)\n",
    "prot2og_withOutgroup = res[0]\n",
    "og2prot_withOutgroup = res[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Proteins "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "prot = pd.read_csv(main_dir + '/data/gbk_parsed/filt/Chlamydia_proteins.txt', delimiter = \"\\t\")\n",
    "prot = prot.rename(index = str, columns = {\"OrthMCL_genome_code\": \"genome_id\", \"OrthMCL_prot_code\": \"protein_id\"})\n",
    "prot.index = prot[\"protein_id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([u'protein_id', u'genome_id', u'product', u'start', u'end', u'strand',\n",
       "       u'length', u'frameshift', u'nucl_seq', u'translation', u'GO_list',\n",
       "       u'EC_list'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. OGroups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4A. only genus Chlamydia (with draft genomes)\n",
    "\n",
    "# column names in OG table (=> to add pfam)\n",
    "col_list_withoutOutgroup = [\"group_id\", \"group_size\", \"distribution\", \"length\", \"all_hypothetical\", \"product\", \"with_frameshift\", \"prod_distr\", \"trachomatis\", \"psittaci\", \"pneumoniae\", \"pecorum\", \"abortus\", \"caviae\", \"felis\", \"muridarum\", \"avium\", \"gallinacea\", \"ibidis\", \"suis\", \"poikilothermis\", \"sanzinia\", \"serpentis\", \"corallus\", \"GOdistr_rast\", \"GO_rast\", \"ECdistr_rast\", \"EC_rast\"]\n",
    "# output data.frame and file\n",
    "ogroups_withoutOutgroup = pd.DataFrame(columns = col_list_withoutOutgroup)\n",
    "outf = main_dir + '/data/ogroups2019/annotated/groups_without_outgroup_with_singletons_annotated.txt'\n",
    "\n",
    "i = 0\n",
    "for group in og2prot_withoutOutgroup.keys():\n",
    "    \n",
    "    # list of proteins in the group\n",
    "    proteins = og2prot_withoutOutgroup[group]\n",
    "    \n",
    "    # !! False if some genomes were excluded after OGs had been created !!\n",
    "    if len(proteins) > 0: \n",
    "        \n",
    "        # prepare\n",
    "        prod_list = []\n",
    "        frame_list = []\n",
    "        genome_list = []\n",
    "        GO_rast = []\n",
    "        EC_rast = []\n",
    "        trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum, avium, gallinacea, ibidis, suis, poikilothermis, sanzinia , serpentis, corallus   = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
    "        product = ''\n",
    "        distribution = 0\n",
    "        \n",
    "        # general statistics\n",
    "        med_length = prot[prot.index.isin(proteins)][\"length\"].median()\n",
    "        OGsize = len(proteins)\n",
    "\n",
    "        # info by protein\n",
    "        for p in proteins:\n",
    "\n",
    "            #products\n",
    "            prod = prot.loc[p,\"product\"]\n",
    "            prod_list.append(prod)\n",
    "\n",
    "            #frameshifts\n",
    "            frame = prot.loc[p, \"frameshift\"]\n",
    "            frame_list.append(frame)\n",
    "\n",
    "            #species and genomes\n",
    "            genome = prot.loc[p, \"genome_id\"]\n",
    "            species = genomes.loc[genome,\"species\"]\n",
    "            exec(species + \"+= 1\")\n",
    "            genome_list.append(genome)\n",
    "\n",
    "            #GO-terms\n",
    "            if not pd.isnull(prot.loc[p, \"GO_list\"]):\n",
    "                GOs = prot.loc[p,\"GO_list\"].split(\";\")\n",
    "                for GO in GOs:\n",
    "                    GO_rast.append(GO)\n",
    "\n",
    "            #EC numbers\n",
    "            if not pd.isnull(prot.loc[p, \"EC_list\"]):\n",
    "                ECs=prot.loc[p,\"EC_list\"].split(\";\")\n",
    "                for EC in ECs:\n",
    "                    EC_rast.append(EC)\n",
    "        \n",
    "        # summary statistics\n",
    "        GOdistr_rast = str(Counter(GO_rast))[8:-1]\n",
    "        GO_rast = ';'.join(str(s) for s in set(GO_rast)) \n",
    "        ECdistr_rast = str(Counter(EC_rast))[8:-1]\n",
    "        EC_rast = ';'.join(str(s) for s in set(EC_rast)) \n",
    "        prod_distr = str(Counter(prod_list))[8:-1]\n",
    "        distribution = len(set(genome_list))\n",
    "        with_frameshift = frame_list.count(\"yes\")\n",
    "        all_hypothetical = 'no'\n",
    "\n",
    "        if all([re.search(\"hypothetic\", m) for m in prod_list]):\n",
    "            product = \"hypothetical protein\"\n",
    "            all_hypothetical = \"yes\"\n",
    "        else:\n",
    "            f = re.match(\"(.+?): [0-9]+\", prod_distr[1:-1]).group(1) # finds the most common product in the group\n",
    "            if re.search(\"hypothetical\", f):\n",
    "                product=\"hypothetical protein\"\n",
    "            else:\n",
    "                product=f[1:-1]    \n",
    "        \n",
    "        l = [group, OGsize, distribution, med_length, all_hypothetical, product, with_frameshift, prod_distr, trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum, avium, gallinacea, ibidis, suis, poikilothermis, sanzinia , serpentis, corallus, GOdistr_rast, GO_rast, ECdistr_rast, EC_rast]\n",
    "        ogroups_withoutOutgroup.loc[i] = l\n",
    "        i += 1\n",
    "\n",
    "ogroups_withoutOutgroup.to_csv(outf, index = False)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4B. genus Chlamydia without draft genomes\n",
    "\n",
    "# column names in OG table (=> to add pfam)\n",
    "col_list_nodraft = [\"group_id\", \"group_size\", \"distribution\", \"trachomatis\", \"psittaci\", \"pneumoniae\", \"pecorum\", \"abortus\", \"caviae\", \"felis\", \"muridarum\", \"avium\", \"gallinacea\", \"ibidis\", \"suis\", \"poikilothermis\", \"sanzinia\", \"serpentis\", \"corallus\"]\n",
    "# output data.frame and file\n",
    "ogroups_nodraft = pd.DataFrame(columns = col_list_nodraft)\n",
    "outf = main_dir + '/data/ogroups2019/annotated/groups_without_outgroup_with_singletons_nodraft_annotated.txt'\n",
    "\n",
    "i = 0\n",
    "for group in og2prot_nodraft.keys():\n",
    "    \n",
    "    # list of proteins in the group\n",
    "    proteins = og2prot_nodraft[group]\n",
    "    \n",
    "    # !! False if some genomes were excluded after OGs had been created !!\n",
    "    if len(proteins) > 0: \n",
    "        \n",
    "        # prepare\n",
    "        genome_list = []\n",
    "        trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum, avium, gallinacea, ibidis, suis, poikilothermis, sanzinia , serpentis, corallus   = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
    "        product = ''\n",
    "        distribution = 0\n",
    "        \n",
    "        # general statistics\n",
    "        OGsize = len(proteins)\n",
    "\n",
    "        # info by protein\n",
    "        for p in proteins:\n",
    "\n",
    "            #species and genomes\n",
    "            genome = prot.loc[p, \"genome_id\"]\n",
    "            species = genomes.loc[genome,\"species\"]\n",
    "            exec(species + \"+= 1\")\n",
    "            genome_list.append(genome)\n",
    "\n",
    "        \n",
    "        # summary statistics\n",
    "        distribution = len(set(genome_list))\n",
    "        with_frameshift = frame_list.count(\"yes\")\n",
    "        \n",
    "        l = [group, OGsize, distribution, trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum, avium, gallinacea, ibidis, suis, poikilothermis, sanzinia , serpentis, corallus]\n",
    "\n",
    "        ogroups_nodraft.loc[i] = l\n",
    "        i += 1\n",
    "\n",
    "ogroups_nodraft.to_csv(outf, index = False)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4C. genus Chlamydia with outgroup\n",
    "\n",
    "# column names in OG table (=> to add pfam)\n",
    "col_list_withOutgroup = [\"group_id\", \"group_size\", \"distribution\", \"trachomatis\", \"psittaci\", \"pneumoniae\", \"pecorum\", \"abortus\", \"caviae\", \"felis\", \"muridarum\", \"avium\", \"gallinacea\", \"ibidis\", \"suis\", \"poikilothermis\", \"sanzinia\", \"serpentis\", \"corallus\", \"Waddlia_chondrophila\"]\n",
    "# output data.frame and file\n",
    "ogroups_withOutgroup = pd.DataFrame(columns = col_list_withOutgroup)\n",
    "outf = main_dir + '/data/ogroups2019/annotated/groups_with_outgroup_with_singletons_annotated.txt'\n",
    "\n",
    "i = 0\n",
    "for group in og2prot_withOutgroup.keys():\n",
    "    \n",
    "    # list of proteins in the group\n",
    "    proteins = og2prot_withOutgroup[group]\n",
    "    \n",
    "    # !! False if some genomes were excluded after OGs had been created !!\n",
    "    if len(proteins) > 0: \n",
    "        \n",
    "        # prepare\n",
    "        genome_list = []\n",
    "        trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum, avium, gallinacea, ibidis, suis, poikilothermis, sanzinia , serpentis, corallus, Waddlia_chondrophila   = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
    "        product = ''\n",
    "        distribution = 0\n",
    "        \n",
    "        # general statistics\n",
    "        OGsize = len(proteins)\n",
    "\n",
    "        # info by protein\n",
    "        for p in proteins:\n",
    "\n",
    "            #species and genomes\n",
    "            genome = prot.loc[p, \"genome_id\"]\n",
    "            species = genomes.loc[genome,\"species\"]\n",
    "            exec(species + \"+= 1\")\n",
    "            genome_list.append(genome)\n",
    "\n",
    "        \n",
    "        # summary statistics\n",
    "        distribution = len(set(genome_list))\n",
    "        with_frameshift = frame_list.count(\"yes\")\n",
    "        \n",
    "        l = [group, OGsize, distribution, trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum, avium, gallinacea, ibidis, suis, poikilothermis, sanzinia , serpentis, corallus, Waddlia_chondrophila]\n",
    "\n",
    "        ogroups_withOutgroup.loc[i] = l\n",
    "        i += 1\n",
    "\n",
    "ogroups_withOutgroup.to_csv(outf, index = False)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['S782', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, '', '', '', '']\n",
      "['S783', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1449', 14, 14, 37.0, 'yes', 'hypothetical protein', 1, \"{'FIG00493827: hypothetical protein': 14}\", 0, 0, 12, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, '', '', '', '']\n",
      "['OG1448', 14, 14, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 14}\", 0, 0, 12, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, '', '', '', '']\n",
      "['OG1339', 20, 20, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 20}\", 0, 0, 0, 0, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1338', 20, 20, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 20}\", 0, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['S784', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['S785', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, '', '', '', '']\n",
      "['OG1335', 21, 21, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 21}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 21, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1334', 21, 21, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 21}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 21, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1337', 20, 20, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 20}\", 0, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1336', 21, 21, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 21}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 21, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1331', 21, 21, 37.0, 'yes', 'hypothetical protein', 3, \"{'FIG00494438: hypothetical protein': 21}\", 0, 0, 11, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1330', 21, 21, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 21}\", 0, 0, 0, 0, 21, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1333', 21, 21, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 21}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 21, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1332', 21, 21, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 21}\", 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 20, 0, 0, 0, 0, '', '', '', '']\n",
      "['S428', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, '', '', '', '']\n",
      "['S355', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1726', 10, 10, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 10}\", 0, 0, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['S761', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, '', '', '', '']\n",
      "['OG1727', 10, 10, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 10}\", 0, 0, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['S541', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1724', 10, 10, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 10}\", 0, 0, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1725', 10, 10, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 10}\", 0, 0, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1201', 29, 29, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 29}\", 0, 0, 0, 0, 29, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG616', 234, 234, 37.0, 'no', \"tRNA (cytidine(34)-2'-O)-methyltransferase (EC 2.1.1.207)\", 0, '{\"tRNA (cytidine(34)-2\\'-O)-methyltransferase (EC 2.1.1.207)\": 234}', 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', \"{'2.1.1.207': 234}\", '2.1.1.207']\n",
      "['OG1203', 29, 29, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 29}\", 0, 0, 0, 0, 29, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1202', 29, 29, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 29}\", 0, 0, 0, 0, 29, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1205', 29, 29, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 29}\", 0, 0, 0, 0, 29, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1204', 29, 29, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 29}\", 0, 0, 0, 0, 29, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1207', 29, 29, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 29}\", 0, 0, 0, 0, 29, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG617', 234, 234, 37.0, 'no', 'FKBP-type peptidyl-prolyl cis-trans isomerase', 27, \"{'FKBP-type peptidyl-prolyl cis-trans isomerase': 132, 'FKBP-type peptidyl-prolyl cis-trans isomerase FklB (EC 5.2.1.8)': 85, 'FKBP-type peptidyl-prolyl cis-trans isomerase / Macrophage infectivity potentiator': 16, 'Peptidylprolyl isomerase, FKBP-type (EC 5.2.1.8)': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', \"{'5.2.1.8': 86}\", '5.2.1.8']\n",
      "['OG1209', 29, 29, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 29}\", 0, 0, 0, 0, 29, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1208', 29, 29, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 29}\", 0, 0, 0, 0, 29, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1422', 15, 15, 37.0, 'yes', 'hypothetical protein', 0, \"{'FIG00494062: hypothetical protein': 12, 'hypothetical protein': 3}\", 0, 0, 12, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, '', '', '', '']\n",
      "['OG614', 234, 234, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 148, 'FIG00493914: hypothetical protein': 86}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG533', 234, 234, 37.0, 'no', 'Arginyl-tRNA synthetase (EC 6.1.1.19)', 1, \"{'Arginyl-tRNA synthetase (EC 6.1.1.19)': 234}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, \"{'GO:0004814': 234}\", 'GO:0004814', \"{'6.1.1.19': 234}\", '6.1.1.19']\n",
      "['OG532', 234, 234, 37.0, 'no', '1-acyl-sn-glycerol-3-phosphate acyltransferase (EC 2.3.1.51)', 1, \"{'1-acyl-sn-glycerol-3-phosphate acyltransferase (EC 2.3.1.51)': 233, 'Acyl-CoA:1-acyl-sn-glycerol-3-phosphate acyltransferase (EC 2.3.1.51)': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, \"{'GO:0003841': 233}\", 'GO:0003841', \"{'2.3.1.51': 234}\", '2.3.1.51']\n",
      "['OG531', 234, 234, 37.0, 'no', 'Cytidylate kinase (EC 2.7.4.25)', 0, \"{'Cytidylate kinase (EC 2.7.4.25)': 234}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', \"{'2.7.4.25': 234}\", '2.7.4.25']\n",
      "['OG530', 234, 234, 37.0, 'no', 'Phosphatidate cytidylyltransferase (EC 2.7.7.41)', 1, \"{'Phosphatidate cytidylyltransferase (EC 2.7.7.41)': 234}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, \"{'GO:0004605': 234}\", 'GO:0004605', \"{'2.7.7.41': 234}\", '2.7.7.41']\n",
      "['OG537', 234, 234, 37.0, 'no', 'GTP-binding protein EngA', 2, \"{'GTP-binding protein EngA': 234}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, \"{'GO:0005622': 234, 'GO:0005525': 234}\", 'GO:0005622;GO:0005525', '', '']\n",
      "['OG536', 234, 234, 37.0, 'no', 'hypothetical protein', 0, \"{'FIG000859: hypothetical protein YebC': 209, 'FIG000859: hypothetical protein': 24, 'Probable transcriptional regulatory protein YebC': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG535', 234, 234, 37.0, 'no', 'Type IV pilus biogenesis protein PilO', 0, \"{'Type IV pilus biogenesis protein PilO': 218, 'FIG01260812: hypothetical protein': 11, 'Predicted cell-wall-anchored protein SasA (LPXTG motif)': 1, 'Streptococcal hemagglutinin protein': 1, 'CT456 hypothetical protein': 1, 'LIPOPROTEIN VSAA (FRAGMENT)': 1, 'hypothetical protein': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, \"{'GO:0009986': 1}\", 'GO:0009986', '', '']\n",
      "['OG534', 234, 234, 37.0, 'no', 'UDP-N-acetylglucosamine 1-carboxyvinyltransferase (EC 2.5.1.7)', 1, \"{'UDP-N-acetylglucosamine 1-carboxyvinyltransferase (EC 2.5.1.7)': 234}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, \"{'GO:0008760': 234}\", 'GO:0008760', \"{'2.5.1.7': 234}\", '2.5.1.7']\n",
      "['S250', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, '', '', '', '']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['OG612', 234, 234, 37.0, 'no', 'DNA polymerase III, epsilon chain', 0, \"{'DNA polymerase III, epsilon chain': 233, 'DNA polymerase III epsilon chain': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG539', 234, 234, 37.0, 'no', 'Peptide chain release factor 2; programmed frameshift-containing', 27, \"{'Peptide chain release factor 2; programmed frameshift-containing': 233, 'Peptide chain release factor 2 @ programmed frameshift-containing': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, \"{'GO:0006415': 234, 'GO:0003747': 234}\", 'GO:0006415;GO:0003747', '', '']\n",
      "['OG538', 234, 234, 37.0, 'no', 'Amino Group Acetyl Transferase', 0, \"{'Amino Group Acetyl Transferase': 234}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['S254', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, '', '', '', '']\n",
      "['S255', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, '', '', '', '']\n",
      "['OG1539', 12, 12, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 12}\", 0, 0, 12, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG613', 234, 234, 37.0, 'no', 'TsaE protein, required for threonylcarbamoyladenosine t(6)A37 formation in tRNA', 0, \"{'TsaE protein, required for threonylcarbamoyladenosine t(6)A37 formation in tRNA': 233, 'tRNA threonylcarbamoyladenosine biosynthesis protein TsaE': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG1029', 68, 68, 37.0, 'yes', 'hypothetical protein', 1, \"{'FIG00493124: hypothetical protein': 67, 'hypothetical protein': 1}\", 67, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1028', 68, 68, 37.0, 'yes', 'hypothetical protein', 6, \"{'FIG00494438: hypothetical protein': 68}\", 0, 25, 12, 0, 29, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, '', '', '', '']\n",
      "['OG610', 234, 234, 37.0, 'no', 'Apolipoprotein N-acyltransferase (EC 2.3.1.-)', 1, \"{'Apolipoprotein N-acyltransferase (EC 2.3.1.-)': 233, 'Apolipoprotein N-acyltransferase': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG1021', 69, 69, 37.0, 'yes', 'hypothetical protein', 1, \"{'hypothetical protein': 53, 'FIG00493479: hypothetical protein': 16}\", 66, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1020', 70, 70, 37.0, 'yes', 'hypothetical protein', 0, \"{'FIG00493256: hypothetical protein': 69, 'hypothetical protein': 1}\", 70, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1023', 69, 69, 37.0, 'yes', 'hypothetical protein', 0, \"{'FIG00493206: hypothetical protein': 68, 'hypothetical protein': 1}\", 68, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG611', 234, 234, 37.0, 'no', 'Cytosolic acyl-CoA thioester hydrolase family protein', 0, \"{'Cytosolic acyl-CoA thioester hydrolase family protein': 234}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG1025', 69, 69, 37.0, 'yes', 'hypothetical protein', 0, \"{'FIG00493194: hypothetical protein': 69}\", 69, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1024', 69, 69, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 69}\", 69, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1027', 68, 68, 37.0, 'no', 'Short-chain dehydrogenase/reductase SDR', 2, \"{'Short-chain dehydrogenase/reductase SDR': 68}\", 0, 25, 0, 10, 29, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, '', '', '', '']\n",
      "['OG1026', 68, 68, 37.0, 'yes', 'hypothetical protein', 0, \"{'FIG00493318: hypothetical protein': 68}\", 65, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG2024', 3, 3, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 3}\", 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['S356', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG227', 234, 234, 37.0, 'no', 'Type III secretion inner membrane protein (YscU,SpaS,EscU,HrcU,SsaU, homologous to flagellar export components)', 0, \"{'Type III secretion inner membrane protein (YscU,SpaS,EscU,HrcU,SsaU, homologous to flagellar export components)': 202, 'Flagellar biosynthesis protein FlhB': 32}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG226', 234, 234, 37.0, 'no', 'Low Calcium Response E (CopN) (Type III secreted protein SctW)', 0, \"{'Low Calcium Response E (CopN) (Type III secreted protein SctW)': 185, 'Inclusion membrane protein-5': 47, 'Scc1 protein (Type III secretion chaperone)': 1, 'Low Calcium Response E (CopN) (Type III secreted protein SctW); Inclusion membrane protein-5': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG225', 234, 234, 37.0, 'no', '4-alpha-glucanotransferase (amylomaltase) (EC 2.4.1.25)', 6, \"{'4-alpha-glucanotransferase (amylomaltase) (EC 2.4.1.25)': 234}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, \"{'GO:0004134': 234}\", 'GO:0004134', \"{'2.4.1.25': 234}\", '2.4.1.25']\n",
      "['OG224', 234, 234, 37.0, 'yes', 'hypothetical protein', 0, \"{'FIG00493859: hypothetical protein': 233, 'hypothetical protein': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG223', 234, 234, 37.0, 'no', 'LSU ribosomal protein L28p', 0, \"{'LSU ribosomal protein L28p': 233, 'LSU ribosomal protein L28p @ LSU ribosomal protein L28p, zinc-independent': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, \"{'GO:0003735': 234, 'GO:0006412': 234, 'GO:0005840': 234, 'GO:0005622': 234}\", 'GO:0003735;GO:0006412;GO:0005840;GO:0005622', '', '']\n",
      "['OG222', 234, 234, 37.0, 'no', 'UbiD family decarboxylase', 0, \"{'UbiD family decarboxylase': 137, 'UbiD family decarboxylase associated with menaquinone via futalosine': 94, '3-polyprenyl-4-hydroxybenzoate carboxy-lyase (EC 4.1.1.-)': 3}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG221', 234, 234, 37.0, 'no', 'Phospholipase D endonuclease superfamily', 1, \"{'Phospholipase D endonuclease superfamily': 234}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG220', 234, 234, 37.0, 'yes', 'hypothetical protein', 1, \"{'FIG00494168: hypothetical protein': 232, 'hypothetical protein': 2}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', '', '']\n",
      "['OG999', 75, 75, 37.0, 'no', 'Inclusion membrane protein-32', 0, \"{'Inclusion membrane protein-32': 27, 'hypothetical protein': 25, 'CT288 hypothetical protein': 23}\", 0, 25, 12, 0, 29, 1, 1, 0, 1, 2, 1, 0, 1, 0, 1, 1, '', '', '', '']\n",
      "['OG998', 76, 76, 37.0, 'yes', 'hypothetical protein', 0, \"{'FIG00493713: hypothetical protein': 76}\", 76, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG229', 234, 234, 37.0, 'no', 'tRNA pseudouridine synthase B (EC 4.2.1.70)', 5, \"{'tRNA pseudouridine synthase B (EC 4.2.1.70)': 232, 'tRNA pseudouridine(55) synthase (EC 5.4.99.25)': 1, 'tRNA pseudouridine synthase B (EC 4.2.1.70) ## tRNA Psi55': 1}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, '', '', \"{'4.2.1.70': 233, '5.4.99.25': 1}\", '4.2.1.70;5.4.99.25']\n",
      "['OG228', 234, 234, 37.0, 'no', 'GTP-binding and nucleic acid-binding protein YchF', 0, \"{'GTP-binding and nucleic acid-binding protein YchF': 234}\", 115, 25, 12, 10, 29, 1, 1, 3, 1, 2, 1, 30, 1, 1, 1, 1, \"{'GO:0005622': 234, 'GO:0005525': 234}\", 'GO:0005622;GO:0005525', '', '']\n",
      "['S874', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1155', 35, 35, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 35}\", 0, 24, 0, 0, 11, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1154', 35, 24, 37.0, 'no', 'helicase, putative', 1, \"{'helicase, putative': 33, 'hypothetical protein': 1, 'Putative helicase': 1}\", 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 32, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1157', 34, 34, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 34}\", 0, 25, 0, 0, 9, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1156', 34, 34, 37.0, 'yes', 'hypothetical protein', 0, \"{'FIG00494493: hypothetical protein': 21, 'hypothetical protein': 13}\", 0, 19, 0, 10, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, '', '', '', '']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['OG1151', 36, 36, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 36}\", 36, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1150', 36, 36, 37.0, 'yes', 'hypothetical protein', 16, \"{'hypothetical protein': 36}\", 0, 5, 0, 0, 29, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, '', '', '', '']\n",
      "['OG1153', 36, 36, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 36}\", 5, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 29, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1152', 36, 36, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 36}\", 34, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1159', 34, 34, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 34}\", 34, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1158', 34, 34, 37.0, 'no', 'phosphatidylcholine-hydrolyzing phospholipase D (PLD) family', 0, \"{'phosphatidylcholine-hydrolyzing phospholipase D (PLD) family': 34}\", 34, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1506', 13, 13, 37.0, 'no', 'CHLPS 43 kDa protein homolog_4', 0, \"{'CHLPS 43 kDa protein homolog_4': 13}\", 0, 0, 12, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, '', '', '', '']\n",
      "['S766', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['S84', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['S388', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1507', 13, 13, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 13}\", 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 12, 0, 0, 0, 0, '', '', '', '']\n",
      "['S389', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['OG1504', 13, 13, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 13}\", 0, 0, 12, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, '', '', '', '']\n",
      "['S812', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['S813', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, '', '', '', '']\n",
      "['S810', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['S811', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n",
      "['S816', 1, 1, 37.0, 'yes', 'hypothetical protein', 0, \"{'hypothetical protein': 1}\", 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', '', '', '']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "        \n",
    "\n",
    "        \n",
    "#### TO BE ADDED #####        \n",
    "#         #pfam info\n",
    "#         if len(proteins)>1:\n",
    "#             query=\"select * from pfam where protein_id in \"+str(tuple(proteins))\n",
    "#         else:\n",
    "#             query=\"select * from pfam where protein_id='\"+str(proteins[0])+\"'\"\n",
    "#         df=pd.read_sql(query, connection)\n",
    "#         pfams=df[\"accession\"].tolist()\n",
    "#         pfam_distr=str(Counter(pfams))[8:-1]\n",
    "#         pfams=';'.join(str(s) for s in set(pfams)) \n",
    "#         #pfam to GO\n",
    "#         GO_pfam=df[\"pfam2go\"].tolist()\n",
    "#         GO_pfam=[x.split(\"|\") for x in GO_pfam]\n",
    "#         GO_pfam=list(itertools.chain.from_iterable(GO_pfam))\n",
    "#         GOdistr_pfam=str(Counter(GO_pfam))[8:-1]\n",
    "#         GO_pfam=';'.join(str(s) for s in set(GO_pfam)) \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# without outgroup, 128 genomes\n",
    "\n",
    "for group in og2prot_withoutOutgroup.keys():\n",
    "    proteins=og2prot_withoutOutgroup[group]\n",
    "    if len(proteins)>0: # excluded proteins (if some genomes are to be excluded after OGs are created)\n",
    "        OGsize=len(proteins)\n",
    "        prod_list=[]\n",
    "        frame_list=[]\n",
    "        genome_list=[]\n",
    "        GO_rast=[]\n",
    "        EC_rast=[]\n",
    "        trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum,avium,gallinacea, ibidis=[0,0,0,0,0,0,0,0,0,0,0]\n",
    "        product=''\n",
    "        distribution=0\n",
    "        avg_length=prot[prot.index.isin(proteins)][\"length\"].mean()\n",
    "\n",
    "\n",
    "        for p in proteins:\n",
    "\n",
    "            #products\n",
    "            prod=prot.loc[p,\"product\"]\n",
    "            prod_list.append(prod)\n",
    "\n",
    "            #frameshifts\n",
    "            frame=prot.loc[p,\"frameshift\"]\n",
    "            frame_list.append(frame)\n",
    "\n",
    "            #species and genomes\n",
    "            genome=prot.loc[p,\"genome_id\"]\n",
    "            species=genomes.loc[genome,\"species\"]\n",
    "            exec(species + \"+= 1\")\n",
    "            genome_list.append(genome)\n",
    "\n",
    "\n",
    "            #GO-terms\n",
    "            if not pd.isnull(prot.loc[p,\"RAST2GO\"]):\n",
    "                GOs=prot.loc[p,\"RAST2GO\"].split(\";\")\n",
    "                for GO in GOs:\n",
    "                    GO_rast.append(GO)\n",
    "\n",
    "\n",
    "            #EC numbers\n",
    "            if not pd.isnull(prot.loc[p,\"RAST2EC\"]):\n",
    "                ECs=prot.loc[p,\"RAST2EC\"].split(\";\")\n",
    "                for EC in ECs:\n",
    "                    EC_rast.append(EC)\n",
    "        \n",
    "        #summary statistics\n",
    "        GOdistr_rast=str(Counter(GO_rast))[8:-1]\n",
    "        GO_rast=';'.join(str(s) for s in set(GO_rast)) \n",
    "        ECdistr_rast=str(Counter(EC_rast))[8:-1]\n",
    "        EC_rast=';'.join(str(s) for s in set(EC_rast)) \n",
    "        prod_distr=str(Counter(prod_list))[8:-1]\n",
    "        distribution=len(set(genome_list))\n",
    "        with_frameshift=frame_list.count(\"yes\")\n",
    "        all_hypothetical='no'\n",
    "\n",
    "        if all([re.search(\"hypothetic\",m) for m in prod_list]):\n",
    "            product=\"hypothetical protein\"\n",
    "            all_hypothetical=\"yes\"\n",
    "        else:\n",
    "            f=re.match(\"(.+?): [0-9]+\",prod_distr[1:-1]).group(1) # finds the most common product in the group\n",
    "            if re.search(\"hypothetical\",f):\n",
    "                product=\"hypothetical protein\"\n",
    "            else:\n",
    "                product=f[1:-1]    \n",
    "        \n",
    "        #pfam info\n",
    "        if len(proteins)>1:\n",
    "            query=\"select * from pfam where protein_id in \"+str(tuple(proteins))\n",
    "        else:\n",
    "            query=\"select * from pfam where protein_id='\"+str(proteins[0])+\"'\"\n",
    "        df=pd.read_sql(query, connection)\n",
    "        pfams=df[\"accession\"].tolist()\n",
    "        pfam_distr=str(Counter(pfams))[8:-1]\n",
    "        pfams=';'.join(str(s) for s in set(pfams)) \n",
    "        #pfam to GO\n",
    "        GO_pfam=df[\"pfam2go\"].tolist()\n",
    "        GO_pfam=[x.split(\"|\") for x in GO_pfam]\n",
    "        GO_pfam=list(itertools.chain.from_iterable(GO_pfam))\n",
    "        GOdistr_pfam=str(Counter(GO_pfam))[8:-1]\n",
    "        GO_pfam=';'.join(str(s) for s in set(GO_pfam)) \n",
    "\n",
    "        #add to database\n",
    "        out_list=[group, OGsize, distribution,avg_length, all_hypothetical,product,with_frameshift,prod_distr,trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum,avium,gallinacea, ibidis, pfam_distr,pfams,GOdistr_pfam,GO_pfam, GOdistr_rast,GO_rast, ECdistr_rast,EC_rast]\n",
    "        cursor.execute(sql_og_withoutOutgroup,(out_list))\n",
    "        #out_list=[str(x) for x in out_list]\n",
    "        #out_list=\"\\t\".join(out_list)+\"\\n\"\n",
    "\n",
    "\n",
    "# with outgroup, 129 genomes\n",
    "\n",
    "for group in og2prot_withOutgroup.keys():\n",
    "    proteins=og2prot_withOutgroup[group]\n",
    "    if len(proteins)>0: # excluded proteins (if some genomes are to be excluded after OGs are created)\n",
    "        OGsize=len(proteins)\n",
    "        prod_list=[]\n",
    "        frame_list=[]\n",
    "        genome_list=[]\n",
    "        GO_rast=[]\n",
    "        EC_rast=[]\n",
    "        trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum, avium, gallinacea, ibidis, Waddlia_chondrophila = [0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "        product=''\n",
    "        distribution=0\n",
    "        avg_length=prot[prot.index.isin(proteins)][\"length\"].mean()\n",
    "\n",
    "\n",
    "        for p in proteins:\n",
    "\n",
    "            #products\n",
    "            prod=prot.loc[p,\"product\"]\n",
    "            prod_list.append(prod)\n",
    "\n",
    "            #frameshifts\n",
    "            frame=prot.loc[p,\"frameshift\"]\n",
    "            frame_list.append(frame)\n",
    "\n",
    "            #species and genomes\n",
    "            genome=prot.loc[p,\"genome_id\"]\n",
    "            species=genomes.loc[genome,\"species\"]\n",
    "            exec(species + \"+= 1\")\n",
    "            genome_list.append(genome)\n",
    "\n",
    "\n",
    "            #GO-terms\n",
    "            if not pd.isnull(prot.loc[p,\"RAST2GO\"]):\n",
    "                GOs=prot.loc[p,\"RAST2GO\"].split(\";\")\n",
    "                for GO in GOs:\n",
    "                    GO_rast.append(GO)\n",
    "\n",
    "\n",
    "            #EC numbers\n",
    "            if not pd.isnull(prot.loc[p,\"RAST2EC\"]):\n",
    "                ECs=prot.loc[p,\"RAST2EC\"].split(\";\")\n",
    "                for EC in ECs:\n",
    "                    EC_rast.append(EC)\n",
    "        \n",
    "        #summary statistics\n",
    "        GOdistr_rast=str(Counter(GO_rast))[8:-1]\n",
    "        GO_rast=';'.join(str(s) for s in set(GO_rast)) \n",
    "        ECdistr_rast=str(Counter(EC_rast))[8:-1]\n",
    "        EC_rast=';'.join(str(s) for s in set(EC_rast)) \n",
    "        prod_distr=str(Counter(prod_list))[8:-1]\n",
    "        distribution=len(set(genome_list))\n",
    "        with_frameshift=frame_list.count(\"yes\")\n",
    "        all_hypothetical='no'\n",
    "\n",
    "        if all([re.search(\"hypothetic\",m) for m in prod_list]):\n",
    "            product=\"hypothetical protein\"\n",
    "            all_hypothetical=\"yes\"\n",
    "        else:\n",
    "            f=re.match(\"(.+?): [0-9]+\",prod_distr[1:-1]).group(1) # finds the most common product in the group\n",
    "            if re.search(\"hypothetical\",f):\n",
    "                product=\"hypothetical protein\"\n",
    "            else:\n",
    "                product=f[1:-1]    \n",
    "        \n",
    "        #pfam info\n",
    "        if len(proteins)>1:\n",
    "            query=\"select * from pfam where protein_id in \"+str(tuple(proteins))\n",
    "        else:\n",
    "            query=\"select * from pfam where protein_id='\"+str(proteins[0])+\"'\"\n",
    "        df=pd.read_sql(query, connection)\n",
    "        pfams=df[\"accession\"].tolist()\n",
    "        pfam_distr=str(Counter(pfams))[8:-1]\n",
    "        pfams=';'.join(str(s) for s in set(pfams)) \n",
    "        #pfam to GO\n",
    "        GO_pfam=df[\"pfam2go\"].tolist()\n",
    "        GO_pfam=[x.split(\"|\") for x in GO_pfam]\n",
    "        GO_pfam=list(itertools.chain.from_iterable(GO_pfam))\n",
    "        GOdistr_pfam=str(Counter(GO_pfam))[8:-1]\n",
    "        GO_pfam=';'.join(str(s) for s in set(GO_pfam)) \n",
    "\n",
    "        #add to database\n",
    "        out_list=[group, OGsize, distribution,avg_length, all_hypothetical,product,with_frameshift,prod_distr,trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum,avium,gallinacea, ibidis,Waddlia_chondrophila, pfam_distr,pfams,GOdistr_pfam,GO_pfam, GOdistr_rast,GO_rast, ECdistr_rast,EC_rast]\n",
    "        cursor.execute(sql_og_withOutgroup,(out_list))\n",
    "        #out_list=[str(x) for x in out_list]\n",
    "        #out_list=\"\\t\".join(out_list)+\"\\n\"\n",
    "\n",
    "\n",
    "# without outgroup, no drafts\n",
    "\n",
    "\n",
    "\n",
    "for group in og2prot_nodraft.keys():\n",
    "    proteins=og2prot_nodraft[group]\n",
    "    if len(proteins)>0: # excluded proteins (if some genomes are to be excluded after OGs are created)\n",
    "        OGsize=len(proteins)\n",
    "        prod_list=[]\n",
    "        frame_list=[]\n",
    "        genome_list=[]\n",
    "        GO_rast=[]\n",
    "        EC_rast=[]\n",
    "        trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum,avium,gallinacea, ibidis, Waddlia_chondrophila = [0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "        product=''\n",
    "        distribution=0\n",
    "        avg_length=prot[prot.index.isin(proteins)][\"length\"].mean()\n",
    "        for p in proteins:\n",
    "            #products\n",
    "            prod=prot.loc[p,\"product\"]\n",
    "            prod_list.append(prod)\n",
    "            #frameshifts\n",
    "            frame=prot.loc[p,\"frameshift\"]\n",
    "            frame_list.append(frame)\n",
    "            #species and genomes\n",
    "            genome=prot.loc[p,\"genome_id\"]\n",
    "            species=genomes.loc[genome,\"species\"]\n",
    "            exec(species + \"+= 1\")\n",
    "            genome_list.append(genome)\n",
    "            #GO-terms\n",
    "            if not pd.isnull(prot.loc[p,\"RAST2GO\"]):\n",
    "                GOs=prot.loc[p,\"RAST2GO\"].split(\";\")\n",
    "                for GO in GOs:\n",
    "                    GO_rast.append(GO)\n",
    "            #EC numbers\n",
    "            if not pd.isnull(prot.loc[p,\"RAST2EC\"]):\n",
    "                ECs=prot.loc[p,\"RAST2EC\"].split(\";\")\n",
    "                for EC in ECs:\n",
    "                    EC_rast.append(EC)\n",
    "        #summary statistics\n",
    "        GOdistr_rast=str(Counter(GO_rast))[8:-1]\n",
    "        GO_rast=';'.join(str(s) for s in set(GO_rast)) \n",
    "        ECdistr_rast=str(Counter(EC_rast))[8:-1]\n",
    "        EC_rast=';'.join(str(s) for s in set(EC_rast)) \n",
    "        prod_distr=str(Counter(prod_list))[8:-1]\n",
    "        distribution=len(set(genome_list))\n",
    "        with_frameshift=frame_list.count(\"yes\")\n",
    "        all_hypothetical='no'\n",
    "        if all([re.search(\"hypothetic\",m) for m in prod_list]):\n",
    "            product=\"hypothetical protein\"\n",
    "            all_hypothetical=\"yes\"\n",
    "        else:\n",
    "            f=re.match(\"(.+?): [0-9]+\",prod_distr[1:-1]).group(1) # finds the most common product in the group\n",
    "            if re.search(\"hypothetical\",f):\n",
    "                product=\"hypothetical protein\"\n",
    "            else:\n",
    "                product=f[1:-1]    \n",
    "        #pfam info\n",
    "        if len(proteins)>1:\n",
    "            query=\"select * from pfam where protein_id in \"+str(tuple(proteins))\n",
    "        else:\n",
    "            query=\"select * from pfam where protein_id='\"+str(proteins[0])+\"'\"\n",
    "        df=pd.read_sql(query, connection)\n",
    "        pfams=df[\"accession\"].tolist()\n",
    "        pfam_distr=str(Counter(pfams))[8:-1]\n",
    "        pfams=';'.join(str(s) for s in set(pfams)) \n",
    "        #pfam to GO\n",
    "        GO_pfam=df[\"pfam2go\"].tolist()\n",
    "        GO_pfam=[x.split(\"|\") for x in GO_pfam]\n",
    "        GO_pfam=list(itertools.chain.from_iterable(GO_pfam))\n",
    "        GOdistr_pfam=str(Counter(GO_pfam))[8:-1]\n",
    "        GO_pfam=';'.join(str(s) for s in set(GO_pfam)) \n",
    "        #add to database\n",
    "        out_list=[group, OGsize, distribution,avg_length, all_hypothetical,product,with_frameshift,prod_distr,trachomatis, psittaci, pneumoniae, pecorum, abortus, caviae, felis, muridarum,avium,gallinacea, ibidis,pfam_distr,pfams,GOdistr_pfam,GO_pfam, GOdistr_rast,GO_rast, ECdistr_rast,EC_rast]\n",
    "        cursor.execute(sql_og_nodraft,(out_list))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# ######### Pfams ##################################################\n",
    "\n",
    "# print (\"parsing pfams info\\n\\n\")\n",
    "\n",
    "# pfam=pd.read_csv(main_dir+\"info/InterProScan_results/Chlamydia_interproscan_w_outgroup_pfam.tsv\",sep=\"\\t\", names=[\"protein_id\", \"Sequence_MD5_digest\", \"length\", \"analysis\", \"accession\", \"description\", \"start\", \"stop\", \"score\", \"status\", \"date\", \"ipr_annotation\", \"ipr_description\", \"GO\"],index_col=False)\n",
    "# ids=set([x.split(\"|\")[1] for x in pfam[\"protein_id\"].tolist() if x.split(\"|\")[0] in genome_ids_withOutgroup])\n",
    "# pfam_dict=dict((zip(ids,[1]*len(ids))))\n",
    "\n",
    "# for i in range (pfam.shape[0]):\n",
    "#     p=pfam.loc[i,\"protein_id\"].split(\"|\")[1]\n",
    "#     try:\n",
    "#         pfam[\"id_short\"]=p\n",
    "#         num=pfam_dict[p]\n",
    "#         pfam_dict[p]+=1\n",
    "#         pfam_id=p+\".\"+str(num)\n",
    "#         acc=pfam.loc[i,\"accession\"]\n",
    "#         desc=pfam.loc[i,\"description\"]\n",
    "#         start=pfam_descr=pfam.loc[i,\"start\"]\n",
    "#         end=pfam.loc[i,\"stop\"]\n",
    "#         pfam2go=pfam.loc[i,\"GO\"]\n",
    "#         if pd.isnull(pfam2go):\n",
    "#             pfam2go=\"\"\n",
    "#         cursor.execute(sql_pfam,(pfam_id,p,acc,desc,start,end,pfam2go))\n",
    "#     except:\n",
    "#         pass\n",
    "\n",
    "#################### Orthologous groups ################################\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
